### ISSUE TITLE
keepalived doesn't run notify_* scripts in order?

### ISSUE THREAD

**Issue by synfinatic** [acassen/keepalived#1055]
I'm not not 100% sure what is going on, but I can see the results.  Basically, keepalived was supposed to run my notify_master script and then the notify_backup script in a very short period of time:

```
$ tail -25 /var/log/keepalived_vrrp.log
Wed Oct 31 19:59:42 2018: Opening file '/etc/keepalived/keepalived.conf'.
Wed Oct 31 19:59:42 2018: Unable to load ipset library - libipset.so.3: cannot open shared object file: No such file or directory
Wed Oct 31 19:59:42 2018: VRRP_Instance(ETH2) removing protocol VIPs.
Wed Oct 31 19:59:42 2018: VRRP_Instance(ETH2) removing protocol iptable drop rule
Wed Oct 31 19:59:42 2018: Using LinkWatch kernel netlink reflector...
Wed Oct 31 19:59:42 2018: VRRP_Instance(ETH2) Entering BACKUP STATE
Wed Oct 31 19:59:42 2018: Opening script file /opt/xxx/keepalived/notify.sh
Wed Oct 31 19:59:42 2018: VRRP sockpool: [ifindex(4), proto(112), unicast(0), fd(11,12)]
Wed Oct 31 19:59:42 2018: VRRP_Script(monitor_ec2vpn) succeeded
Wed Oct 31 20:15:50 2018: VRRP_Instance(ETH2) Transition to MASTER STATE
Wed Oct 31 20:15:51 2018: VRRP_Instance(ETH2) Entering MASTER STATE
Wed Oct 31 20:15:51 2018: VRRP_Instance(ETH2) setting protocol iptable drop rule
Wed Oct 31 20:15:51 2018: VRRP_Instance(ETH2) setting protocol VIPs.
Wed Oct 31 20:15:51 2018: Sending gratuitous ARP on eth2 for 192.168.60.1
Wed Oct 31 20:15:51 2018: VRRP_Instance(ETH2) Sending/queueing gratuitous ARPs on eth2 for 192.168.60.1
Wed Oct 31 20:15:51 2018: Sending gratuitous ARP on eth2 for 192.168.60.1
Wed Oct 31 20:15:51 2018: Sending gratuitous ARP on eth2 for 192.168.60.1
Wed Oct 31 20:15:51 2018: Sending gratuitous ARP on eth2 for 192.168.60.1
Wed Oct 31 20:15:51 2018: Sending gratuitous ARP on eth2 for 192.168.60.1
Wed Oct 31 20:15:51 2018: Opening script file /opt/xxx/keepalived/notify.sh
Wed Oct 31 20:15:51 2018: VRRP_Instance(ETH2) Master received advert with higher priority 200, ours 100
Wed Oct 31 20:15:51 2018: VRRP_Instance(ETH2) Entering BACKUP STATE
Wed Oct 31 20:15:51 2018: VRRP_Instance(ETH2) removing protocol VIPs.
Wed Oct 31 20:15:51 2018: VRRP_Instance(ETH2) removing protocol iptable drop rule
Wed Oct 31 20:15:51 2018: Opening script file /opt/xxx/keepalived/notify.sh
```

my keepalived.conf:
```
cat /etc/keepalived/keepalived.conf
global_defs {
    router_id VPN_RTR_2_1
    vrrp_skip_check_adv_addr
    vrrp_strict
    vrrp_mcast_group4 224.0.0.18
    vrrp_skip_check_adv_addr false
    vrrp_garp_master_delay 5
    vrrp_garp_master_repeat 5
    vrrp_garp_master_refresh 0
    vrrp_garp_master_refresh_repeat 1
    vrrp_no_swap
    checker_no_swap
    enable_script_security
    script_user root root
}

vrrp_script monitor_ec2vpn {
    # make sure ec2vpn is open for business
    script "/opt/xxx/keepalived/vpn_status.sh"
    # check every X seconds
    interval 15
    # require X failures for KO
    fall 2
    # require X successes for OK
    rise 2
}

vrrp_instance ETH2 {
    state BACKUP
    interface eth2
    virtual_router_id 1
    priority 100
    advert_int 1
    virtual_ipaddress {
        192.168.60.1
    }
    track_script {
        monitor_ec2vpn
    }
    notify_master "/opt/xxx/keepalived/notify.sh master"
    notify_backup "/opt/xxx/keepalived/notify.sh backup"
    notify_fault "/opt/xxx/keepalived/notify.sh fault"
}
```

But the result is that either the backup ran before the master or the backup didn't run.  Which means my external systems end up thinking this node is the 'master' even though keepalived knows it is really the 'backup'.

I've verified that traffic is flowing to/through the other/master system and this system really is the backup.

My notify.sh:
```
#!/bin/bash
# Script to be called by keepalived when notify events happen

if [ -z "$1" ]; then
    echo "Error: Please provide an action"
    exit 1
fi

STATUS_FILE=/run/keepalived.status

case $1 in
    master)
        echo "master" >$STATUS_FILE
        ;;
    backup)
        echo "backup" >$STATUS_FILE
        ;;
    fault)
        echo "fault" >$STATUS_FILE
        ;;
    *)
        echo "UKNOWN STATUS: $1" >$STATUS_FILE
        exit 1
        ;;
esac

exit 0
```

**Comment by synfinatic** [comment]
Almost forgot:
```
$ keepalived -v
Keepalived v1.4.2 (02/24,2018), git commit v1.4.1-41-g6a2987e+

Copyright(C) 2001-2018 Alexandre Cassen, <acassen@gmail.com>

Built with kernel headers for Linux 3.10.0
Running on Linux 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017

Build options:  PIPE2 LIBNL3 RTA_ENCAP RTA_EXPIRES FRA_OIFNAME FRA_TUN_ID RTAX_CC_ALGO RTAX_QUICKACK LIBIPTC LIBIPSET_DYNAMIC NET_LINUX_IF_H_COLLISION LVS LIBIPVS_NETLINK VRRP VRRP_AUTH VRRP_VMAC SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE OLD_CHKSUM_COMPAT FIB_ROUTING INET6_ADDR_GEN_MODE SNMP_V3_FOR_V2 SNMP SNMP_VRRP SNMP_CHECKER SNMP_RFC SNMP_RFCV2 SNMP_RFCV3 SO_MARK
```

**Comment by pqarmitage** [comment]
This is a known issue, and is an inevitable consequence of notify scripts being executed as soon as the triggering event occurs. It has been suggested that keepalived should wait for the execution of one notify script before it runs the next, but on a system with a large number of state transitions this isn't feasible.

It was for this reason that I implemented `notify_fifo` etc, where the order of delivery of messages on the FIFO is guaranteed to be correct; it is also much more efficient than executing scripts on every state transition. Keepalived can also run a user defined script to process the messages received on the FIFO, configured via `notify_fifo_script` etc. `doc/samples/sample_notify_fifo.sh` is a sample script to process reading messages from the FIFO.

The `notidy_fifo` feature is available in v1.4.2.

**Comment by synfinatic** [comment]
thanks!
